### ISSUE TITLE
Proposal: Enhance docker info output with volume driver list

### ISSUE THREAD

**Issue by jimmyxian** [docker/docker#16205]
Currently the output of `docker info` does not have infomation about volume drivers.
When create a volume using `--driver=`, I want to know what driver are supported. But No where to get.

Maybe we can add volume driver list in docker info, such as:
```
[root@Host80 ~]# docker info
Containers: 1
Images: 64
Storage Driver: devicemapper
...
Volume Drivers: local,a,b,c
...
Name: Host80
```

Also, this way we can also use it in Swarm. Swarm can get supported volume driver list on every machines. Then `create volume` in swarm can use it(see related issue on the Swarm side docker/swarm#1189).

**Comment by jimmyxian** [comment]
ping @vieux @thaJeztah @abronan @aluzzardi @chanwit

**Comment by thaJeztah** [comment]
Yes, would be good to have this info available somewhere.

/cc @albers - could be useful for the completion scripts :-)

**Comment by albers** [comment]
@thaJeztah yes, I would use this functionality in bash completion.

**Comment by phemmer** [comment]
Putting this information in `docker info` requires that the docker daemon be running. However the available storage drivers is something you would want to know before starting docker, as if you start docker with one storage driver, and you want to switch, you'll likely need to wipe out `/var/lib/docker`.

**Comment by thaJeztah** [comment]
@phemmer sorry for the confusion; this is about the "Volume drivers" ("volume plugins"), although a list of available graph/storage drivers would be nice as well (i.e. `aufs`, `btrfs`, `overlay`)

I wonder if we need the same for "network plugins" / drivers? If we do, is it preferred to have these drivers/plugins listed at the "top" level, or nested? e.g.

```
Plugins:
    Volume Drivers: a, b, c, d
    Network Drivers: e, f, g, h
```

**Comment by jimmyxian** [comment]
@thaJeztah 
Yeah, nested is more clear.

**Comment by albers** [comment]
The nested structure makes the output harder to parse, though.
For bash completion I need to extract the information with shell tools, which work best when the relevant information is on one line, i.e. 
```
Volume Driver plugins: a, b, c, d
Network Driver plugins: e, f, g, h
```
I admit that this can be done with awk, but I don't like the idea of introducing the concept of structured information in this place. With this being in place, future additions might introduce even deeper nested structures, and mix different kinds of structuring in one place.

**Comment by albers** [comment]
I just noticed that the output of `docker info` already contains structured information:
```
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 90
 Dirperm1 Supported: false
```
So my argument against @thaJeztah's proposed structure is obsolete.

**Comment by thaJeztah** [comment]
For completion, perhaps we should have a `docker plugins` command, e.g. `docker plugins --filter type=volume-driver`. But let's keep that for another issue :relaxed:

**Comment by Pensu** [comment]
@thaJeztah Weren't we having a discussion regarding putting flags in docker info, so that it doesn't get much cluttered?

**Comment by thaJeztah** [comment]
@Pensu true, and that's still a hard one to solve (what's verbose for one person's needs could be "required information" for another). I think the information talked about here (available plugins) is useful for Swarm, so I'm +1 for adding it. The "nested" presentation could be used in future to make the output more customizable (e.g. do/do not show the "plugins" section)

For now, I'm fine either way, at risk of breaking the completion scripts when we decide to change it to a struct later :smile:

**Comment by Pensu** [comment]
@thaJeztah Haha, makes sense actually! I am all in for customizable output/ nested structure. Would like to see that happening, as we are going to need a lot more info in swarm when the scheduler becomes more mature.

**Comment by abronan** [comment]
Sounds good to me ;)

This will soon be needed as we include the support for volume drivers in Swarm. I guess it makes sense too for a single docker daemon.

The nested structure as proposed by @thaJeztah makes sense to me and is easily readable (and not necessarily harder to parse overall). `info` already has this nested structure for storage drivers.

**Comment by thaJeztah** [comment]
@abronan @cpuguy83 do we want/need this for the 1.9 release?

**Comment by cpuguy83** [comment]
I don't think it is necessary for 1.9, certainly a nice to have.
But it would also be helpful to have a plugin API as well.

**Comment by duglin** [comment]
@diogomonica @NathanMcCauley 
just a little off-topic, but I was wondering if exposing too much information via `docker info` to end users could be a security concern because it might give people some hints as to what security holes to try to exploit. I'm not suggesting we don't do this PR, or others like it, but rather should we consider giving the docker daemon the ability to either turn off `docker info` entirely or to at least modify it to only expose a subset of the information?  Just wondering...

**Comment by thaJeztah** [comment]
@duglin perhaps that's something to consider using the auth plugins for?

**Comment by albers** [comment]
@thaJeztah Support for volume and network plugins was added in #17300, support for authorization pluging was added in #18836.
So this can be closed now.
